Tutorial on building REST APIs with EdgeDB & FastAPI (#3762)
edgedb/edgedb: docs/guides/tutorials/rest_apis_with_fastapi.rst

.. _ref_guide_rest_apis_with_fastapi:

===============================
REST APIs with EdgeDB & FastAPI
===============================

> EdgeDB can help you quickly build REST APIs in Python without getting into the rigmarole of using ORM libraries to handle your data effectively. Here, we'll be using `FastAPI <https://fastapi.tiangolo.com/>`_ to expose the API endpoints and EdgeDB to store the content.

Reviewer: One nit: we limit line lengths to 79 characters max for all files in this repo.

Author: Done

We'll build a simple event management system where you'll be able to fetch, create, update, and delete *event hosts* and *events* via RESTful API endpoints.

Prerequisites
=============

> Before we start, make sure you've :ref:`installed <ref_admin_install>` EdgeDB and EdgeDB-CLI. In this tutorial, we'll use Python 3.10 and take advantage of the asynchronous I/O paradigm to communicate with the database more efficiently. A working version of this tutorial can be found `here <https://github.com/edgedb/edgedb-examples/tree/main/fastapi-crud>`_ on GitHub.

Reviewer: ```suggestion
Before we start, make sure you've :ref:`installed <ref_admin_install>` the `edgedb` command line tool. In this tutorial, we'll use Python 3.10 and take advantage of the asynchronous I/O paradigm to communicate with the database more efficiently. A working version of this tutorial can be found `on GitHub <https://github.com/edgedb/edgedb-examples/tree/main/fastapi-crud>`_.
```

Install the dependencies
^^^^^^^^^^^^^^^^^^^^^^^^

> To follow along, clone the repository and head over to the ``fastapi-crud`` directory. Create a Python 3.10 virtual environment, activate it, and install the dependencies with this command:
> 
> .. code-block:: bash
>     pip install edgedb fastapi httpx[cli] uvicorn

Reviewer: Should we clarify things here? Not everybody is fluent with venvs:

```suggestion
To follow along do the following steps: 

* clone the repository and head over to the ``fastapi-crud`` directory;

* create a Python 3.10 virtual environment and activate it:

  .. code-block:: bash
  
     $ python3.10 -m venv .venv
     $ source .venv/bin/activate

* install the dependencies with this command:
  
  .. code-block:: bash
  
     $ pip install edgedb fastapi httpx[cli] uvicorn

.. code-block:: bash
    pip install edgedb fastapi httpx[cli] uvicorn
```

OTOH giving a name to the venv is sort of confusing... Maybe it's better for the user to figure this out. Or maybe we just link "create a venv" to some popular tutorial or to Python docs?

Thoughts?

Author: Creating venv can be a pain on different OSes. Also, different people prefer different tools to do that like—the built-in venv, virtualenv, conda, etc. 

So, I wanted to avoid this completely. But you're correct. I'll at least add a link here.

Initialize the database
^^^^^^^^^^^^^^^^^^^^^^^

> Now, let's create a new EdgeDB instance for this project. From the root directory, run:

Reviewer: ```suggestion
Now, let's create a new EdgeDB instance for this project. From the project's root directory, run:
```

> .. code-block:: bash
>     edgedb project init

Reviewer: I think you're supposed to have an empty line between ReST directives (`. code-block:: bash` in this case) and its body (`edgedb project init` in this case), like so:

```suggestion
.. code-block:: bash

    edgedb project init
```

Author: I was able to build the site locally. These should be resolved.

You should see the following prompts on your console:

::
    Initializing project...

    Specify the name of EdgeDB instance to use with this project
    [default: fastapi_crud]:
    > fastapi_crud

    Do you want to start instance automatically on login? [y/n]
    > n
    Checking EdgeDB versions...

Once you've answered the prompts, a new EdgeDB instance called ``fastapi_crud`` will be created. Start the server with the following command:

.. code-block:: bash

>     edgedb instance start fastapi_crud

Reviewer: We're about to change `edgedb project init` to always keep the instance running (even when you answered "n" to "Do you want to start instance automatically on login? [y/n]")  (https://github.com/edgedb/edgedb-cli/issues/710)

cc @tailhook @colinhacks @elprans 

But maybe for the sake of this tutorial and to avoid complications *now* we just suggest to answer "y" and then we can avoid this dance with 'edgedb instance start'?

Reviewer 2: Sure. I think `y` should be the answer. If we don't think that auto-start instance is the most common use case we should just drop this question.

Author: I thought the instance keeps running in the background—just like docker containers—when someone logs out. If not, then auto initialization is something that I'd want.

Connect to the database
^^^^^^^^^^^^^^^^^^^^^^^

Let's test that we can connect to the newly started instance. To do so, run:

.. code-block:: bash
    edgedb instance start fastapi_crud

You should be connected to the database instance and able to see a prompt similar to this:

::
    EdgeDB 1.2+5aecabc (repl 1.1.1+5bb8bad)
    Type \help for help, \quit to quit.
    edgedb>

You can start writing queries here. However, we haven't talked about the shape of our data yet.

Schema design
=============

The event management system will have two entities—**events** and **users**. Each *event* can have an optional link to a *user*. The goal is to create API endpoints that'll allow us to fetch, create, update, and delete the entities while maintaining their relationships.

> EdgeDB allows us to declaratively define the structure of the entities. If you've worked with SQLAlchemy or Django ORM, you might refer to these declarative schema definitions as *models*. By default, EdgeDB CLI looks for these schema definitions in the `dbschema/default.esdl` module. You can also create new modules in this directory and call the types from one module to another. However, for now, we'll define our entities in the `default.esdl` module. This is how our datatypes look:

Reviewer: ```suggestion
EdgeDB allows us to declaratively define the structure of the entities. If you've worked with SQLAlchemy or Django ORM, you might refer to these declarative schema definitions as *models*. In EdgeDB we call them "object types". 

By default, EdgeDB CLI looks for these schema definitions in the `dbschema/default.esdl` module. You can also create new modules in this directory and call the types from one module to another. However, for now, we'll define our entities in the `default.esdl` module. This is how our datatypes look:
```

Author: Object types it is!

.. code-block:: esdl
    # dbschema/default.esdl

    module default {

>     abstract type AuditLog {

Reviewer: I think using nouns for mixins makes it a bit harder to understand the inheritance tree. Maybe call it `Auditable`?

Author: Agreed. Changed to auditable.

>       annotation description := "Add 'create_at' and 'update_at' properties to all types.";

Reviewer: ```suggestion
      annotation description := 
          "A mixin type that adds 'create_at' and \
          'update_at' properties to all types.";
```

Author: Yeah. I removed the annotations in the tutorial but kept them around in the example application. Curious readers will find their way.

Author: AFAIK `updated_at` requires Trigger, which isn't supported as of now.

>       property created_at -> datetime {

Reviewer: Should also be "readonly"

User APIs
^^^^^^^^^^

Since the ``User`` type is simpler among the two, we'll start with that. Let's create a `GET users/` endpoint so that we can start looking at the user objects saved in the database. You can create the API with a couple of lines of code in FastAPI:

.. code-block:: python
    # fastapi-crud/app/users.py

    from __future__ import annotations

    import datetime
    from http import HTTPStatus
    from typing import Iterable

    import edgedb
    from fastapi import APIRouter, HTTPException, Query
    from pydantic import BaseModel

    router = APIRouter()
    client = edgedb.create_async_client()

    class RequestData(BaseModel):
        name: str

    class ResponseData(BaseModel):
        name: str
        created_at: datetime.datetime

    @router.get("/users")
    async def get_users(
        name: str = Query(None, max_length=50)
        ) -> Iterable[ResponseData]:

        if not name:
            users = await client.query("SELECT User {name, created_at};")
        else:
            users = await client.query(
                """SELECT User {name, created_at} FILTER User.name=<str>$name""",
                name=name,
            )

>         response = (
>             ResponseData(name=user.name,
>             created_at=user.created_at) for user in users
>         )

Reviewer: This formatting style is a bit easier to read.

```suggestion
        response = (
            ResponseData(
                name=user.name,
                created_at=user.created_at
            ) for user in users
        )
```

It's rather unfortunate that we have to do this repack manually. We need to improve the situation here asap, by providing a way to generate these pydantic models automatically for queries.

.. code-block:: bash
    uvicorn app.main:fast_api --port 5000 --reload

This will start a ``uvicorn`` server and you'll be able to start making requests against it. Earlier, we installed the `HTTPx <https://www.python-httpx.org/>`_ client library to make HTTP requests programmatically. It also comes with a neat command-line tool that we'll use to test our API. While the ``uvicorn`` server is running, on a new console, run:

.. code-block:: bash
    httpx -m GET http://localhost:5000/users

You'll see the following output on the console:

::
    HTTP/1.1 200 OK
    date: Sat, 16 Apr 2022 22:58:11 GMT
    server: uvicorn
    content-length: 2
    content-type: application/json

    []

Our request yielded an empty list because the database doesn't have any object at this point. Let's create the ``POST /users`` endpoint to start saving users in the database. The POST endpoint can be built similarly:

.. code-block:: python
    # fastapi-crud/app/users.py

    ...
    @router.post("/users", status_code=HTTPStatus.CREATED)
    async def post_user(user: RequestData) -> ResponseData:

        try:
            (created_user,) = await client.query(

>                 """SELECT (INSERT User {name:=<str>$name}) {name, created_at};""",

Reviewer: ```suggestion
                """
                WITH 
                    new_user := (INSERT User {name := <str>$name})
                SELECT new_user {
                    name, 
                    created_at
                };""",
```

    {
      "name": "Resuscitation",
      "address": "Britain",
      "schedule": "1889-07-28T06:59:59+00:00",
      "host": {
        "name": "Mina Murray"
      }
    }

You can also use the ``GET /events`` endpoint to list and filter the event objects. To locate the ``Resuscitation`` event, you'd use the ``filter_name`` parameter with the GET API as follows:

.. code-block:: bash
    httpx -m GET http://localhost:5000/events -p 'name' 'Resuscitation'

That'll return:

::
    HTTP/1.1 200 OK
    ...
    {
      "name": "Resuscitation",
      "address": "Britain",
      "schedule": "1889-07-28T06:59:59+00:00",
      "host": {
        "name": "Mina Murray"
      }
    }

Take a look at the ``app/events.py`` file to see how the ``PUT /events`` and ``DELETE /events`` endpoints are constructed.

Browse the endpoints using the native OpenAPI doc
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

FastAPI automatically generates OpenAPI schema from the API endpoints and uses those to build the API docs. While the ``uvicorn`` server is running, go to your browser and head over to `http://localhost:5000/docs <http://locahost:5000/docs>`_. You should see an API navigator like this:

> .. image:: https://user-images.githubusercontent.com/30027932/163834402-1bb766d0-a2c4-4fdf-8b0b-9af2ff47a969.png

Reviewer: We can add this image files to our docs

Author: Wouldn't that blow up the repo size? Do we want that?

Reviewer: Not significantly, but you're right.

Here's a solution I came up with:

* I created a new branch `pages` for this repo and enabled GitHub pages.
* https://github.com/edgedb/edgedb/tree/pages is it.
* We can add our images (and whatever else) there and serve them from GitHub, for example: https://edgedb.github.io/edgedb/images/guides/fastapi/fastapi1.png

Reviewer: cc @colinhacks @jaclarke @vpetrovykh @elprans

Reviewer: Oh well, I'm an idiot -- a separate branch for GH pages is still in the same repo so I achieved nothing. :/

Let's host aux images on our website?
